PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/localrun/FunctionRunState.java

```diff
@@ -44,24 +72,144 @@ public FunctionRunState(@NotNull Project project, FunctionRunConfiguration funct
         this.functionRunConfiguration = functionRunConfiguration;
     }
 
+
     @Override
     protected String getDeployTarget() {
         return "null";
     }
 
+    private void launchDebugger(final Project project, int debugPort) {
+        final Runnable runnable = () -> {
+            final RunManagerImpl manager = new RunManagerImpl(project);
+            final ConfigurationFactory configFactory = RemoteConfigurationType.getInstance().getConfigurationFactories()[0];
+            final RemoteConfiguration remoteConfig = new RemoteConfiguration(project, configFactory);
+            remoteConfig.PORT = String.valueOf(debugPort);
+            remoteConfig.HOST = "localhost";
+            remoteConfig.USE_SOCKET_TRANSPORT = true;
+            remoteConfig.SERVER_MODE = false;
+            remoteConfig.setName("azure functions");
+
+            final RunnerAndConfigurationSettingsImpl configuration = new RunnerAndConfigurationSettingsImpl(manager, remoteConfig, false);
+            manager.setTemporaryConfiguration(configuration);
+            ExecutionUtil.runConfiguration(configuration, ExecutorRegistry.getInstance().getExecutorById(ToolWindowId.DEBUG));
+        };
+        ApplicationManager.getApplication().invokeAndWait(runnable, ModalityState.any());
+    }
+
     @Override
     protected FunctionApp executeSteps(@NotNull RunProcessHandler processHandler, @NotNull Map<String, String> telemetryMap) throws Exception {
-        // Todo: implement run function locally
+        // Prepare staging Folder
+        final File stagingFolder = new File(functionRunConfiguration.getStagingFolder());
+        prepareStagingFolder(stagingFolder, processHandler);
+        // Run Function Host
+        runFunctionCli(processHandler, stagingFolder);
         return null;
     }
 
+    private void runFunctionCli(RunProcessHandler processHandler, File stagingFolder) throws IOException, InterruptedException {
+        final int funcPort = findFreePortForApi();
+        final int debugPort = findFreePortForApi();
+        processHandler.println(String.format("Func host start at port : %s", funcPort), ProcessOutputTypes.SYSTEM);
+        process = getRunFunctionCliProcessBuilder(stagingFolder, funcPort, debugPort).start();
+        // Add listener to close func.exe
+        addProcessTerminatedListener(processHandler, process);
+        // Redirect function cli output to console
+        processProcessInputStream(process.getInputStream(), string -> {
+            if (isDebugMode() && StringUtils.containsIgnoreCase(string, "Job host started")) {
+                // launch debugger when func ready
+                launchDebugger(project, debugPort);
+            }
+            processHandler.setText(string);
+        });
+        processProcessInputStream(process.getErrorStream(), string -> processHandler.println(string, ProcessOutputTypes.STDERR));
```

Comment:
Have you tested the case when processHandler is destroyed and  `processHandler.println` for last piece of message.

Reply:
Added check for process, will only print if process is alive